File: 2001-jump-game-vii/jump-game-vii.py
class Solution:
    def canReach(self, s: str, minJump: int, maxJump: int) -> bool:
        n = len(s)
        dp = [False]*n
        dp[0]=True

        reachableCount = 0
        for i in range(1,n):
            addIndex = i-minJump
            if (addIndex>=0 and dp[addIndex]==True):
                reachableCount+=1
            
            removeIndex = i-maxJump-1
            if (removeIndex>=0 and dp[removeIndex]==True):
                reachableCount-=1
            
            if (s[i]=='0' and reachableCount>0):
                dp[i]=True

        return dp[-1]

File: 2001-jump-game-vii/jump-game-vii.cpp
class Solution {
public:
    bool canReach(string s, int minJump, int maxJump) {
        int n = s.size();
        vector<bool> dp(n, false);
        dp[0] = true;
        int reachableCount = 0;

        for (int i = 1; i < n; ++i) {
            int addIndex = i - minJump;
            if (addIndex >= 0 && dp[addIndex]) {
                reachableCount++;
            }

            int removeIndex = i - maxJump - 1;
            if (removeIndex >= 0 && dp[removeIndex]) {
                reachableCount--;
            }

            if (s[i] == '0' && reachableCount > 0)
                dp[i] = true;
        }
        return dp[n - 1];
    }
};